### PR TITLE
add `lorri watch --once`

### PR DESCRIPTION
**Pull request by Profpatsch**
Adds a `--once` argument to lorri watch, to only run one evaluation.

`lorri install` was proposed instead, but we don’t want it to sound
like the command changes the system of the user (think `make
install`).

Ultimately, `lorri watch` is going to be replaced by `lorri daemon`
completely, then we should introduce a `lorri run` command which does
the same as `lorri watch --once`.

Co-Authored-By: Graham Christensen <graham@grahamc.com>
Co-Authored-By: Tobias Pflug <tobias.pflug@tweag.io>

### CODE REVIEW

File: src/ops/mod.rs [hunk 1]
```diff
@@ -55,7 +55,7 @@ impl ExitError {
     /// Helpers to create exit results
     ///
     /// Note: err panics if exitcode is zero.
-    fn err<T>(exitcode: i32, message: T) -> ExitError
+    pub fn err<T>(exitcode: i32, message: T) -> ExitError
```
Comment (grahamc) [inline src/ops/mod.rs]: Let's make a function which creates an error of exitcode 100, with a variable message, named to represent what that error code represents.

File: src/ops/watch.rs [hunk 2]
```diff
@@ -1,28 +1,42 @@
...
+                build_loop.forever(tx);
+            })
+        };
 
-    for msg in rx {
-        println!("{:#?}", msg);
-    }
+        for msg in rx {
+            println!("{:#?}", msg);
+            let _ = std::io::stdout().flush();
```
Comment (grahamc) [inline src/ops/watch.rs]: Tell me about this?

File: src/ops/watch.rs [hunk 3]
```diff
@@ -1,28 +1,42 @@
...
     let (tx, rx) = channel();
 
-    let build_thread = {
-        thread::spawn(move || {
-            let mut build_loop = BuildLoop::new(&project);
-            build_loop.forever(tx);
-        })
-    };
+    if opts.once {
+        let mut build_loop = BuildLoop::new(&project);
```
Comment (grahamc) [inline src/ops/watch.rs]: Let's move the `once` codepath in to a function (inside this same file) called `main_run_once`, and move the `forever` codepath in to a function (again, same file) called `main_run_forever`, and switch between them here in `main.

File: src/ops/mod.rs (hunk 1 above)
Comment (Profpatsch) [reply src/ops/mod.rs]: Not actually using this in this commit, leftover from splitting out of the other PR probably. Will remember it when the merge conflicts hit.

File: src/ops/watch.rs (hunk 2 above)
Comment (Profpatsch) [reply src/ops/watch.rs]: Split into different commit.

**Comment by Profpatsch** [comment]
@grahamc Addressed suggestions, please see force push and new commits.

File: src/ops/watch.rs [hunk 4]
```diff
@@ -19,10 +43,19 @@ pub fn main(project: Project) -> OpResult {
...
+        print_build_message(msg);
+        let _ = std::io::stdout().flush();
     }
 
     build_thread.join().unwrap();
 
     ok()
 }
+
+// TODO: this should probably be something more structured
```
Comment (grahamc) [inline src/ops/watch.rs]: Let's drop the TODO. We'll know it needs to be TO-DONE when we get to the feature:

```suggestion
```

File: src/ops/watch.rs [hunk 5]
```diff
@@ -19,10 +43,19 @@ pub fn main(project: Project) -> OpResult {
     };
 
     for msg in rx {
-        println!("{:#?}", msg);
+        print_build_message(msg);
+        let _ = std::io::stdout().flush();
```
Comment (grahamc) [inline src/ops/watch.rs]:
```suggestion
        let _ = std::io::stdout().flush();
```

Should this be part of `print_build_message`?

**Pull request merged**